### PR TITLE
rgw/s3select: update s3select submodule to last commit, (new features), update for test coverage(s3test)

### PR DESCRIPTION
**Pull request by galsalomon66**
Signed-off-by: gal salomon <gal.salomon@gmail.com>

## Checklist
- [ ] References tracker ticket
- [x] Updates documentation if necessary
- [x] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by galsalomon66** [comment]
48  passed, none failed

http://pulpito.front.sepia.ceph.com/gsalomon-2021-04-20_00:58:27-rgw:verify-update_s3select_submodule_and_s3tests_head_10apr-distro-basic-smithi/

**Comment by galsalomon66** [comment]
running teuthology with change of compare sign == -> =

2 failures (46 passed), not related to s3select 

http://pulpito.front.sepia.ceph.com/gsalomon-2021-05-05_11:07:28-rgw:verify-update_s3select_submodule_and_s3tests_head_10apr-distro-basic-smithi/

File: doc/radosgw/s3select.rst [hunk 1]
```diff
@@ -69,11 +69,11 @@ Features Support
...
-|                                 | ``%`` modulo    | select count(*) from s3object where cast(_1 as int)%2 == 0;           |
+|                                 | ``%`` modulo    | select count(*) from s3object where cast(_1 as int)%2 = 0;            |
 +---------------------------------+-----------------+-----------------------------------------------------------------------+
 |                                 | ``^`` power-of  | select cast(2^10 as int) from s3object;                               |
 +---------------------------------+-----------------+-----------------------------------------------------------------------+
-| Compare operators               | > < >= <= == != | select _1,_2 from s3object where (int(_1)+int(_3))>int(_5);           |
+| Compare operators               | > < >= <= = !=  | select _1,_2 from s3object where (int(_1)+int(_3))>int(_5);           |
 +---------------------------------+-----------------+-----------------------------------------------------------------------+
 | logical operator                | AND OR NOT      | select count(*) from s3object where not (int(1)>123 and int(_5)<200); |
```
Comment (mattbenjamin) [inline doc/radosgw/s3select.rst:78]: what is the difference between "int(_5)" and int(5)" (i.e., wrt the underscore);  I think I see these intermixed

Comment (mattbenjamin) [reply doc/radosgw/s3select.rst:78]: I've never seen _<number> in a sql dialect before--is that something aws s3select actually does?

Comment (mattbenjamin) [reply doc/radosgw/s3select.rst:78]: I assume these must be column names;  Is this ansi sql?  I've only seem aliases used for this

Comment (galsalomon66) [reply doc/radosgw/s3select.rst:78]: > what is the difference between "int(_5)" and int(5)" (i.e., wrt the underscore); I think I see these intermixed

it's a mistake, it should be _{column positions)

**Comment by galsalomon66** [comment]
jenkins test make check
jenkins test docs
jenkins test api

**Comment by galsalomon66** [comment]
all had passed 

http://pulpito.front.sepia.ceph.com/gsalomon-2021-05-11_11:55:05-rgw:verify-update_s3select_submodule_and_s3tests_head_10apr-distro-basic-smithi/

**Comment by galsalomon66** [comment]
jenkins test make check

**Review by mattbenjamin — APPROVED** [review]
lgtm

File: doc/radosgw/s3select.rst [hunk 2]
```diff
@@ -267,6 +361,27 @@ Alias
...
+    
+    | s3select contains several testing frameworks which provide a large coverage for its functionalities.
+
+    | (1) tests comparison against trusted engine, meaning,  C/C++ compiler is a trusted expression evaluator, 
+    | since the syntax for arithmetical and logical expressions are identical (s3select compare to C) 
+    | the framework runs equal expressions and validates their results.
+    | A dedicated expression generator produces different sets of expressions per each new test session. 
+
+    | (2) compare results of queries whose syntax is different but semantically they are equal.
+    | this kind of test validates that different runtime flows produces identical result,
```
[the remark this answers is not in the document]
Comment (galsalomon66) [reply doc/radosgw/s3select.rst]: :+1:

File: doc/radosgw/s3select.rst [hunk 3]
```diff
@@ -215,20 +247,80 @@ s3-select function interfaces
 
 Timestamp functions
 ~~~~~~~~~~~~~~~~~~~
-    | The `timestamp functionalities <https://docs.aws.amazon.com/AmazonS3/latest/dev/s3-glacier-select-sql-reference-date.html>`_ is partially implemented.
-    | the casting operator( ``timestamp( string )`` ), converts string to timestamp basic type.
-    | Currently it can convert the following pattern ``yyyy:mm:dd hh:mi:dd``
+    | The `timestamp functionalities <https://docs.aws.amazon.com/AmazonS3/latest/dev/s3-glacier-select-sql-reference-date.html>`_ is fully implemented.
```
[the remark this answers is not in the document]
Comment (galsalomon66) [reply doc/radosgw/s3select.rst]: :+1:

File: doc/radosgw/s3select.rst [hunk 4]
```diff
@@ -269,11 +269,11 @@ Timestamp functions
     | supported date-part : year,month,week,day,hour,minute,second,timezone_hour,timezone_minute.
 
     | ``dateadd(date-part, integer, timestamp)`` : function return timestamp, a calculation results of input timestamp and date-part.
-    | supported data-part : year,month,day,day,hour,minute,second.
+    | supported data-part : year,month,day,hour,minute,second.
     | dateadd operator includes timezone in calculation.
 
     | ``datediff(date-part, timestamp, timestamp)`` : function return an integer, a calculated result for difference between 2 timestamps according to date-part.
```
[the remark this answers is not in the document]
Comment (galsalomon66) [reply doc/radosgw/s3select.rst]: :+1:

**Comment by tchaikov** [comment]
could you prefix the title of your commit message with the subcomponent your are changing ? see https://github.com/ceph/ceph/blob/master/SubmittingPatches.rst#3-describe-your-changes

**Comment by mattbenjamin** [comment]
> 
> 
> could you prefix the title of your commit message with the subcomponent your are changing ? see https://github.com/ceph/ceph/blob/master/SubmittingPatches.rst#3-describe-your-changes

so this could be rgw/s3select?

Matt

**Comment by tchaikov** [comment]
@galsalomon66 please note, the title of the commit message is not identical to that of a PR. you might want to use 
```
git commit --amend
```
to revise your commit message for updating it.

File: qa/suites/rgw/verify/tasks/s3tests.yaml [hunk 5]
```diff
@@ -1,5 +1,5 @@
 tasks:
 - s3tests:
     client.0:
-      force-branch: ceph-master
+      force-branch: tethology_s3select_apr19
```
Comment (cbodley) [inline qa/suites/rgw/verify/tasks/s3tests.yaml:4]: oops! this part shouldn't have merged

Comment (cbodley) [reply qa/suites/rgw/verify/tasks/s3tests.yaml:4]: is there anything on this `tethology_s3select_apr19` branch that should be merged into `ceph-master`?

Comment (galsalomon66) [reply qa/suites/rgw/verify/tasks/s3tests.yaml:4]: no. that should not be merged into ceph-master

**Pull request merged**